Adding an extra option to suppress rate metrics if so desired

hasSoumya opened the pull request:
Please take a look. This will save consumers $$. We are trying to weed away with rate metrics as well - since there's ways in librato to do this without client side data.

Basic idea is to allow the users to use a new factory function to pass in specific rate metrics they want instead of giving them no option. A little more democratic :)

hasSoumya commented:
@mihasya Would love for you to take a look and let me know if this is something we can integrate.

mihasya commented:
Hey that certainly sounds useful.. I have had this thought before.

High level feedback - this patch is pretty hard to read; it is not very clear exactly what's happening and why the code achieves the end. Specifically:
- Need better-named methods. `handleDefaultRateOptions` and `handleRateOption` seem like they should do nearly identical things, but they're completely different methods. `handle` as a "catch-all" verb is hurting readability.
- I'm not sure I understand the `RateOptions` data structure. It seems like the values are always `nil`. Are you just using the "map as a set" pattern here? Perhaps naming could make this better as well - `EnabledRates` or `EnabledRateSet` perhaps.
- I don't like the use of `reflect` inside the reporter loop. I understand what you're trying to do in using a DRY method for both the cases, but reflection has a cost, both in terms of processing and readability. I believe `Timer` implements `Meter` so you could achieve the same end by just type asserting `Timer` as `Meter` and making your DRY method work with an instance of `Meter`.

Let me know if these make sense. Thanks!

hasSoumya commented:
Hi, to answer your comments:
- Yes - the naming of the methods could be clearer, I'm fixing that.
- The rate data structure holds specifically which options I want to turn on say, Rate1 for Timer but Rate15 for Meter. And yes, this is a Hash Set so to speak - because I didn't want the client to repeat the values. I can name this differently as well.
- Your thoughts on reflection are well taken, but my understanding of this piece of code was that it would get invoked every X seconds lowest being 5 - I wouldn't imagine those reflect calls having a big overhead there. 

I'm fairly new to Golang - so keep those comments coming. I'll make the changes and update this PR this morning at some point.

hasSoumya commented:
@mihasya I've updated some names and code a little bit. On the comment about reflection - I generally agree with you - but this piece of code will be called a maximum of 6 times per run - which at the lowest resolution will be around every 5 seconds in it own goroutine. If you still feel strongly about it - I can refactor the code - however, I don't believe my DRY philosophy is actually affecting performance in any significant manner. I added some more tests, and fixed a glaring bug with the rate method not being invoked. 

I can also add a map of functions to their types - but that's probably over engineering it. Point is that the original code was a little too repetitive, whichever way we choose to go, we should move forward with something more compact. Thanks. Looking forward to your comments.

mihasya commented:
(on my plate to review this weekend, so sorry for the long wait. have had my hands full)

hasSoumya commented:
No problems. I do have a small update for this - I will update the PR shortly.

mihasya commented on librato.go:
```diff
@@ -21,18 +22,40 @@ func translateTimerAttributes(d time.Duration) (attrs map[string]interface{}) {
+	return &Reporter{e, t, s, d, r, p, translateTimerAttributes(u), nil, int64(d / time.Second)}
+}
+
+func NewReporterWithRateOptions(r metrics.Registry, o EnableRateSet, d time.Duration, e string, t string, s string, p []float64, u time.Duration) *Reporter {
```
I personally prefer the pattern of adding one method of the following form per one optional member. Otherwise, you end up with a combinatoric explosion of constructors every time you add a new optional thing: 

``` go
func (self *Reporter) WithXType(someX XType) {
    self.x = someX;
    return self
}

# Usage
foo := NewReporter(/* necessary args */).
    WithXType(instanceOfXType).
    WithYType(instanceOfYType).
    WithZType(instanceOfZType)
```

Then your code 

If you can switch that up, I'll happily merge, rest looks great

hasSoumya replied:
Thanks, I'll make the change shortly. I have been slacking on paying attention to this note.
